### PR TITLE
gpu: Build and Ship an GPU enabled Kernel

### PR DESCRIPTION
**Pull request by zvonkok**
With each release make sure we ship a GPU and TEE enabled kernel

Fixes: https://github.com/kata-containers/kata-containers/issues/6553

DependsOn: ~~https://github.com/kata-containers/kata-containers/pull/6522~~

### CODE REVIEW

File: tools/packaging/kernel/build-kernel.sh [hunk 1]
```diff
@@ -415,10 +416,37 @@ build_kernel() {
...
+build_kernel_headers() {
+	echo "### BUILD KERNEL HEADERS"
+	local kernel_path=${1:-}
+	[ -n "${kernel_path}" ] || die "kernel_path not provided"
+	[ -d "${kernel_path}" ] || die "path to kernel does not exist, use ${script_name} setup"
+	[ -n "${arch_target}" ] || arch_target="$(uname -m)"
+	arch_target=$(arch_to_kernel "${arch_target}")
+	pushd "${kernel_path}" >>/dev/null
+
+	if [ "$linux_headers" == "deb" ]; then
```
Comment (jepio) [inline tools/packaging/kernel/build-kernel.sh:428]: CI is failing with `/home/runner/work/kata-containers/kata-containers/tools/packaging/kernel/build-kernel.sh: line 428: linux_headers: unbound variable`.

File: tools/packaging/kernel/build-kernel.sh [hunk 2]
```diff
@@ -415,10 +416,37 @@ build_kernel() {
...
+	if [ "$linux_headers" == "rpm" ]; then
+		make -j $(nproc ${CI:+--ignore 1}) rpm-pkg ARCH="${arch_target}"
+	fi
+
+	popd >>/dev/null
+}
+
+install_kernel_headers() {
+	echo "### INSTALL KERNEL HEADERS"
+
```
Comment (jepio) [inline tools/packaging/kernel/build-kernel.sh:440]: What's the TODO here?

**Comment by jepio** [comment]
What's the design here - where is the build of the actual kernel modules?

**Comment by jepio** [comment]
The CI is also requesting that `tools/packaging/kernel/kata_config_version` get bumped.

File: tools/packaging/kernel/build-kernel.sh (hunk 1 above)
Comment (zvonkok) [reply tools/packaging/kernel/build-kernel.sh:428]: This should be fixed now.

File: tools/packaging/kernel/build-kernel.sh (hunk 2 above)
Comment (zvonkok) [reply tools/packaging/kernel/build-kernel.sh]: Oops, forgot to clean up.

**Comment by zvonkok** [comment]
@jepio One missing part is the caching of the linux-headers, which is going to be used in the rootfs-nvidia-build PR coming soon, this is first to have the right kernels ready.

**Comment by zvonkok** [comment]
@jepio I am automating some of the steps illustrated here: https://github.com/kata-containers/kata-containers/blob/main/docs/use-cases/NVIDIA-GPU-passthrough-and-Kata.md

**Comment by zvonkok** [comment]
If `conf_guest` is set we need to update the CONFIG_LOCALVERSION to match the suffix created in install_kata
 `-nvidia-gpu-{snp|tdx}`, the linux headers will be named the very same `-nvidia-gpu` if build with make deb-pkg for TDX or SNP. The headers may get overwritten if TDX and SNP use the very same kernel version.

**Comment by zvonkok** [comment]
With the new naming the `rootf` has now the correct `linux-headers` 
```
root@whataever:/usr/src# ls -l
total 12
drwxr-xr-x. 6 root root 4096 Apr  5 09:00 linux-headers-5.19.0-nvidia-gpu-tdx
drwxr-xr-x. 6 root root 4096 Apr  5 09:00 linux-headers-5.19.2-nvidia-gpu-snp
drwxr-xr-x. 8 root root 4096 Apr  5 07:12 nvidia-525.89.02
```

**Comment by jepio** [comment]
Could you try to keep things in individual commits as much as possible. Changing LOCALVERSION handling, adding a new TDX kernel, etc. all that can be a single commit per change.

I assume you will build the GPU modules during rootfs build time, and the final artifact will have neither kernel headers nor GPU installer in it?

**Comment by zvonkok** [comment]
Yes, it will be a clean rootfs without any build artifacts.

**Comment by zvonkok** [comment]
Let me get the commits "straight", static-checks were complaining that the commit was malformed when I had several commits.

**Comment by zvonkok** [comment]
@jepio @fidencio PTAL

**Comment by zvonkok** [comment]
@jepio @fidencio PTAL tested on a "clean" system.

**Comment by zvonkok** [comment]
❯ sudo ctr run --runtime "io.containerd.kata.v2" --rm -t "$image" test-kata uname -r
5.19.2-nvidia-gpu-snp

File: tools/packaging/kernel/build-kernel.sh [hunk 3]
```diff
@@ -487,10 +510,12 @@ install_kata() {
 	ls -la "${install_path}/vmlinux${suffix}.container"
 	ls -la "${install_path}/vmlinuz${suffix}.container"
 	popd >>/dev/null
+
+	set +x
```
Comment (amshinde) [inline tools/packaging/kernel/build-kernel.sh:514]: Is this a debug statement left behind?

File: tools/packaging/kernel/build-kernel.sh [hunk 4]
```diff
@@ -487,10 +510,12 @@ install_kata() {
...
 }
 
 main() {
-	while getopts "a:b:c:deEfg:hk:p:t:u:v:x:" opt; do
+	while getopts "a:b:c:deEfg:hH:k:p:t:u:v:x:" opt; do
```
Comment (amshinde) [inline tools/packaging/kernel/build-kernel.sh:518]: Can you add this new option added to show when usage() is invoked?

**Comment by amshinde** [comment]
@zvonkok Could you also include adding necessary documentation changes to `packaging/kernel/README.md` to reflect the changes you are introducing here?

File: .github/workflows/build-kata-static-tarball-amd64.yaml [hunk 5]
```diff
@@ -21,6 +21,8 @@ jobs:
           - kernel
           - kernel-dragonball-experimental
           - kernel-tdx-experimental
+          - kernel-gpu-snp
+          - kernel-gpu-tdx
```
Comment (fidencio) [inline .github/workflows/build-kata-static-tarball-amd64.yaml:25]: I like it.

There are two things that make me a little bit concerned, though.
1) We don't ship a SNP kernel yet
2) I'd prefer renaming `kernel-gpu-tdx` `kernel-gpu-tdx-experimental`, as the TDX kernel is not yet upstreamed (and the same applies for SNP, I think).
3) I'd also love to see the GPU being provided for the "vanilla" kernel, but this one can be a follow-up PR.

File: .github/workflows/release.yaml [hunk 6]
```diff
@@ -92,7 +92,7 @@ jobs:
           tarball="kata-containers-$tag-vendor.tar.gz"
           pushd $GITHUB_WORKSPACE
           bash -c "tools/packaging/release/generate_vendor.sh ${tarball}"
-          GITHUB_TOKEN=${{ secrets.GIT_UPLOAD_TOKEN }} hub release edit -m "" -a "${tarball}" "${tag}" 
+          GITHUB_TOKEN=${{ secrets.GIT_UPLOAD_TOKEN }} hub release edit -m "" -a "${tarball}" "${tag}"
```
Comment (fidencio) [inline .github/workflows/release.yaml:95]: Not related to this PR. :-)

File: tools/packaging/kata-deploy/local-build/Makefile [hunk 7]
```diff
@@ -26,6 +26,8 @@ all: serial-targets \
 	kernel-tarball \
 	kernel-dragonball-experimental-tarball \
 	kernel-tdx-experimental-tarball \
+	kernel-gpu-snp-tarball \
+	kernel-gpu-tdx-tarball \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/Makefile:30]: Same comment I made about the "experimental" bits, and also about having this for the "vanilla" kernel.

File: tools/packaging/kernel/build-kernel.sh [hunk 8]
```diff
@@ -545,7 +588,7 @@ main() {
 		if [ -n "$kernel_version" ];  then
 			kernel_major_version=$(get_major_kernel_version "${kernel_version}")
 			if [[ ${kernel_major_version} != "5.10" ]]; then
-				info "dragonball-experimental kernel patches are only tested on 5.10.x kernel now, other kernel version may cause confliction"	
+				info "dragonball-experimental kernel patches are only tested on 5.10.x kernel now, other kernel version may cause confliction"
```
Comment (fidencio) [inline tools/packaging/kernel/build-kernel.sh:591]: Changes not related to this PR. :-)

File: tools/packaging/static-build/kernel/Dockerfile [hunk 9]
```diff
@@ -2,7 +2,7 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
-FROM ubuntu:20.04
+FROM ubuntu:22.04
```
Comment (fidencio) [inline tools/packaging/static-build/kernel/Dockerfile:5]: Out of curiosity, why do we need to bump to 22.04?

**Review by fidencio — COMMENTED** [review]
@zvonkok, I've left a few comments.

This is a very good PR, and I'll help you work out the caching part for those.

File: .github/workflows/build-kata-static-tarball-amd64.yaml (hunk 5 above)
Comment (zvonkok) [reply .github/workflows/build-kata-static-tarball-amd64.yaml:25]: We ship a SNP guest-kernel the 5.19.2 its in the version.yaml. Vanilla yep make sense.

File: .github/workflows/release.yaml (hunk 6 above)
Comment (zvonkok) [reply .github/workflows/release.yaml:95]: Oh ding it.

File: tools/packaging/static-build/kernel/Dockerfile (hunk 9 above)
Comment (zvonkok) [reply tools/packaging/static-build/kernel/Dockerfile:5]: I am building a 22.04 rootfs, kernel and modules need to be built with the same kernel. Only 22.04 has the open-source drivers shipped.

Comment (zvonkok) [reply tools/packaging/static-build/kernel/Dockerfile:5]: We should anticipate switching all the builds to 22.04 aka jammy.

File: tools/packaging/kernel/build-kernel.sh (hunk 3 above)
Comment (zvonkok) [reply tools/packaging/kernel/build-kernel.sh:514]: Indeed, great catch :)

File: tools/packaging/kernel/build-kernel.sh (hunk 8 above)
Comment (zvonkok) [reply tools/packaging/kernel/build-kernel.sh:591]: Meh, done.

File: tools/packaging/static-build/kernel/Dockerfile (hunk 9 above)
Comment (fidencio) [reply tools/packaging/static-build/kernel/Dockerfile:5]: Yeah, I have no issues about that, that was just to understand the why. :-)

**Review by fidencio — APPROVED** [review]
lgtm, thanks @zvonkok!

Once this one gets merged I will add the caching jobs for those.

**Comment by zvonkok** [comment]
@amshinde PTAL thanks.

**Review by amshinde — APPROVED** [review]

**Pull request merged**